File: cpp/chapter2/lake_counting.cpp
/* 適当なwからはじめ、繋がっているwを.に置き換えていくように繰り返す。
 * mainからdfsを呼び出した数が答えになる。
 */

#include <iostream>
#include <vector>

using namespace std;

vector<vector<char> > field;
int n, m;

void dfs(int x, int y){
  field[x][y] = '.';

  for(int dx=-1; dx <= 1; dx++){
    for(int dy=-1; dy <= 1; dy++){
      int next_x = x + dx;
      int next_y = y + dy;

      if(0 <= next_x && next_x < n && 0 <= next_y && next_y < m &&
          field[next_x][next_y] == 'w'){
        dfs(next_x, next_y);
      }
    }
  }
  return;
}

int main(void){
  cin >> n >> m;
  field = vector<vector<char> >(n, vector<char>(m, 0));
  for(int i=0; i < field.size(); i++){
    for(int j=0; j < field[i].size(); j++){
      cin >> field[i][j];
    }
  }

  int lake_num = 0;
  for(int i=0; i < field.size(); i++){
    for(int j=0; j < field[i].size(); j++){
      if(field[i][j] == 'w'){
        dfs(i, j);
        lake_num++;
      }
    }
  }

  cout << lake_num << endl;

  return 0;
}

File: python/chapter2/lake_counting.py
"""
N＊Mの庭に、水たまりができた。
水たまりは８近傍で隣接している。
全部で幾つの水たまりがあるか。

10 12
W........WW.
.WWW.....WWW
....WW...WW.
.........WW.
.........W..
..W......W..
.W.W.....WW.
W.W.W.....W.
.W.W......W.
..W.......W.
->3
"""

def dfs(x, y):
    """
    深さ優先探索で(x,y)から'W'を探して'.'に置き換えられるまで繰り返す。

    Parameters
    ----------
    x : int
        現在位置x座標
    y : int
        現在位置y座標
    """
    field[i][j] = '.'

    for dx in [-1, 0, 1]:
        for dy in [-1, 0, 1]:
            nx = x + dx
            ny = y + dy
            if 0 <= nx < n and 0 <= ny < m and field[nx][ny] == 'W':
                dfs(nx, ny)

n, m = map(int, input().split())
field = list()
for _ in range(n):
    field.append(list(input()))

lake_num = 0
for i in range(n):
    for j in range(m):
        if (field[i][j] == 'W'):
            dfs(i, j)
            lake_num += 1

print(lake_num)
